File: java/src/main/java/fr/sdecout/handson/rest/book/BookController.java
package fr.sdecout.handson.rest.book;

import fr.sdecout.handson.rest.author.AuthorId;
import fr.sdecout.handson.rest.shared.Isbn;
import fr.sdecout.handson.rest.shared.ValidIsbn;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.stream.Stream;

import static org.springframework.http.HttpStatus.NO_CONTENT;
import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;

@RestController
@RequestMapping("/books")
public class BookController {

    private final BookAccess bookAccess;
    private final BookSearch bookSearch;
    private final BookUpdate bookUpdate;

    BookController(
            BookAccess bookAccess,
            BookSearch bookSearch,
            BookUpdate bookUpdate
    ) {
        this.bookAccess = bookAccess;
        this.bookSearch = bookSearch;
        this.bookUpdate = bookUpdate;
    }

    @GetMapping(value = "/{isbn}", produces = APPLICATION_JSON_VALUE)
    public ResponseEntity<BookResponse> getBook(@ValidIsbn @PathVariable String isbn) {
        return bookAccess.findBook(new Isbn(isbn))
                .map(ResponseEntity::ok)
                .orElseGet(ResponseEntity.notFound()::build);
    }

    @PostMapping(value = "/search", produces = APPLICATION_JSON_VALUE)
    public Stream<BookSearchResponseItem> searchBooks(@NotBlank @RequestParam String hint) {
        return bookSearch.searchBooks(hint);
    }

    @PutMapping(value = "/{isbn}", consumes = APPLICATION_JSON_VALUE)
    @ResponseStatus(NO_CONTENT)
    public void addOrUpdateBook(@ValidIsbn @PathVariable String isbn, @Valid @RequestBody BookUpdateRequest requestBody) {
        bookUpdate.save(
                new Isbn(isbn),
                requestBody.title(),
                requestBody.authors().stream().map(AuthorId::new).toList()
        );
    }

}

File: kotlin/src/main/kotlin/fr/sdecout/handson/rest/book/BookController.kt
package fr.sdecout.handson.rest.book

import fr.sdecout.handson.rest.author.AuthorId
import fr.sdecout.handson.rest.shared.ValidIsbn
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import org.springframework.http.HttpStatus.NO_CONTENT
import org.springframework.http.MediaType.APPLICATION_JSON_VALUE
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/books")
class BookController(
    private val bookAccess: BookAccess,
    private val bookSearch: BookSearch,
    private val bookUpdate: BookUpdate,
) {

    @GetMapping("/{isbn}", produces = [APPLICATION_JSON_VALUE])
    fun getBook(@ValidIsbn @PathVariable isbn: String): ResponseEntity<BookResponse> = bookAccess.findBook(Isbn(isbn))
        ?.let { ResponseEntity.ok(it) }
        ?: ResponseEntity.notFound().build()

    @PostMapping("/search", produces = [APPLICATION_JSON_VALUE])
    fun searchBooks(@NotBlank @RequestParam hint: String): List<BookSearchResponseItem> = bookSearch.searchBooks(hint)

    @PutMapping("/{isbn}", consumes = [APPLICATION_JSON_VALUE])
    @ResponseStatus(NO_CONTENT)
    fun addOrUpdateBook(@ValidIsbn @PathVariable isbn: String, @Valid @RequestBody requestBody: BookUpdateRequest) {
        bookUpdate.save(Isbn(isbn), requestBody.title!!, requestBody.authors!!.map { AuthorId(it) })
    }

}
